src/components/lib/stubs/arch/x86_64/cos_asm_stubs.h

```diff
@@ -102,29 +147,206 @@ __cosrt_s_##name:				\
  *
  * The __cosrt_extern_* aliases are to support components to call a
  * server function in an interface that it also exports.
+ * 
+ * __cosrt_fast_callgate_* is the trampoline for making user-level
+ * synchronous invocations. It does the following:
+ * 	- save callee saved registers to the client stack
+ * 	- save the parameters in rcx and rdx; we will need these 
+ * 		for scratch registers
+ * 	- load the client AUTH token into r15; this value is JITed
+ * 		at boot time
+ * 	- get the tid off the stack
+ * 	- use the tid to index into the user-level kernel memory
+ * 		and get a pointer to this thread's user-level 
+ * 		invocation stack
+ * 	- switch to the user-level kernel's protection domain
+ * 	- push the call's invocation stack from
+ * 		- the client stack
+ * 		- the sinvcap for kernel usage
+ * 	- switch to the server's protection domain
+ * 	- put the invocation token into rbp, just like in 
+ * 		a normal sinv
+ * 	- check the client AUTH token to verify we got without
+ * 		tampering
+ * 	- save the return address into rcx; we are switching stacks
+ * 		and cant using the stack for control flow
+ * 	- jump to the server side of the callgate. this:
+ * 		- gets the server's stack 
+ * 		- sets up the server's stack with the
+ * 			invocation token and thdid
+ * 		- push the return address we put in rcx
```

Comment:
@evanstella @gparmer If the server finds this memory address and somehow changed its value, does this mean that it could return to a random address in client?

Reply:
Yes, just you can randomly mess with your call stack to jump to a random address. This would bypass the return path which switches back to the client's mpk protection domain, leading to a page fault.